src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,500 @@
+#pragma once
+#include <vector>
+#include "common/container/concurrent_blocking_queue.h"
+#include "common/container/concurrent_map.h"
+#include "storage/projected_columns.h"
+#include "storage/projected_row.h"
+#include "storage/write_ahead_log/log_io.h"
+
+namespace terrier::storage {
+// TODO(Zhaozhe): Should fetch block_size. Use magic number first.
+// The block size to output to disk every time
+#define CHECKPOINT_BLOCK_SIZE 4096
+// TODO(Yuning): Should allow setting number of pre-allocated buffers.
+// Use magic number before we have a setting manager.
+#define CHECKPOINT_BUF_NUM 10
+
+// checkpoint file header flags
+#define P_IS_CATALOG 0
+
+/**
+ * The header of a page in the checkpoint file.
+ */
+// TODO(Zhaozhe, Mengyang): More fields can be added to header
+class PACKED CheckpointFilePage {
+ public:
+  /**
+   * Initialize a given page
+   * @param page to be initialized.
+   */
+  static void Initialize(CheckpointFilePage *page) {
+    // TODO(mengyang): support non-trivial initialization
+    page->checksum_ = 0;
+    page->table_oid_ = 0;
+    page->version_ = 0;
+    page->flags_ = 0;
+  }
+
+  /**
+   * Get the checksum of the page.
+   * @return checksum of the page.
+   */
+  uint32_t GetChecksum() { return checksum_; }
+
+  /**
+   * Set the checksum of the page.
+   * @param checksum
+   */
+  void SetCheckSum(uint32_t checksum) { checksum_ = checksum; }
+
+  /**
+   * Get the oid of the table whose rows are stored in this page.
+   * @return oid of the table.
+   */
+  catalog::table_oid_t GetTableOid() { return catalog::table_oid_t(table_oid_); }
+
+  /**
+   * Set the oid of the table whose rows are stored in this page.
+   * @param oid of the table.
+   */
+  void SetTableOid(catalog::table_oid_t oid) { table_oid_ = !oid; }
+
+  /**
+   * Get the pointer to the first row in this page.
+   * @return pointer to the first row.
+   */
+  byte *GetPayload() { return payload_; }
+
+  /**
+   * Set all the flags at the same time.
+   * @param flags to be set.
+   */
+  void SetFlags(uint32_t flags) { flags_ = flags; }
```

Comment:
There is not mention of flags in your design doc images (https://drive.google.com/file/d/16bLiW8dKNqBdfS7nP2iAUIGgGKKBESPp/view). It might help to edit them to include the flags.